`findextrema`: compute `findmin` and `findmax` in single pass

andrewjradcliffe opened the pull request:
This is a simple extension of extant `findmin` and `findmax` methods. Depending on context (cost of `f`; whether reduction is over dims; size of array) the speedup increase is somewhere between 1.0-1.6 (no regressions). Interestingly, I noticed but could not locate a `findextrema`; there is some [mention](https://github.com/JuliaLang/julia/pull/7327) of it, but nothing in Base. If it was deemed unworthy, please excuse this errant PR.

andrewjradcliffe committed:
`findextrema`: compute `findmin` and `findmax` in single pass

This is a simple extension of extant `findmin` and `findmax` methods.
Depending on context (cost of `f`; whether reduction is over dims;
size of array) the speedup increase is somewhere between 1.0-1.6 (no regressions).
Interestingly, I noticed but could not locate a `findextrema`; there
is some [mention](https://github.com/JuliaLang/julia/pull/7327) of it,
but nothing in Base. If it was deemed unworthy, please excuse this
errant PR.

andrewjradcliffe committed:
squash! `findextrema`: compute `findmin` and `findmax` in single pass

andrewjradcliffe committed:
squash! squash! `findextrema`: compute `findmin` and `findmax` in single pass

mikmoore commented:
`extrema` is to `(minimum,maximum)` as `minmax` is to `(min,max)`, so perhaps `findminmax` is the a more consistent name here? Although there isn't an obvious notion of a distinction between `findminmax` and `findextrema`, so either would be appropriate. Wait to see what other people feel about this before you bother to change it.

StefanKarpinski commented:
Thanks for the PR—this does seem like desirable functionality! Anyone familiar with the reduce machinery willing to code review this?

andrewjradcliffe committed:
Improve performance -- single `f` evaluation

Embarrassingly, the original `findextrema` over all elements did not
guarantee a single evaluation of `f` per iteration.

N5N3 commented on base/reducedim.jl:
```diff
@@ -1231,6 +1231,150 @@ end
+function findextrema!(f, op_mn, op_mx, Rval_mn, Rind_mn, Rval_mx, Rind_mx, A::AbstractArray{T,N}) where {T,N}
+    (isempty(Rval_mn) || isempty(Rval_mx) || isempty(A)) && return ((Rval_mn, Rind_mn), (Rval_mx, Rind_mx))
+    lsiz_mn = check_reducedims(Rval_mn, A)
+    lsiz_mx = check_reducedims(Rval_mx, A)
```
This seems redundant. (We check `axes(Rval_mx) == axes(Rval_mn)` below.)

andrewjradcliffe replied:
There is indeed some redundancy. At most, `check_reducedims` on either `Rval_mn` or `Rval_mx` will suffice.

N5N3 commented on base/reduce.jl:
```diff
@@ -971,6 +971,57 @@ julia> findmin([1, 7, 7, NaN])
+"""
+findextrema(f, domain) = _findextrema(f, domain, :)
+_findextrema(f, domain, ::Colon) = mapfoldl(((k, v),) -> (fv = f(v); ((fv, k), (fv, k))), _rf_findextrema, pairs(domain))
+_rf_findextrema((((fm₁, im₁), (fx₁, ix₁))), (((fm₂, im₂), (fx₂, ix₂)))) =
+    (isgreater(fm₁, fm₂) ? (fm₂, im₂) : (fm₁, im₁)), (isless(fx₁, fx₂) ? (fx₂, ix₂) : (fx₁, ix₁))
```
We can fuse `rf_findmin/max` here:
```julia
_rf_findextrema((a, b), (c, d)) = _rf_findmin(a, c), _rf_findmax(b, d)
```

N5N3 commented on base/reducedim.jl:
```diff
@@ -1231,6 +1231,150 @@ end
 
 reducedim1(R, A) = length(axes1(R)) == 1
 
+function findextrema!(f, op_mn, op_mx, Rval_mn, Rind_mn, Rval_mx, Rind_mx, A::AbstractArray{T,N}) where {T,N}
```
IIUC, this is not for public usage. Then we'd better add a `_` to mark this as an internal method.

andrewjradcliffe replied:
Correct. `_` prefix added.

N5N3 reviewed (commented):
I agree with @mikmoore on the name's consistency.
As the added funtion return `findmin(), findmax()` rather than `extrema(), indexes`.
Although `findminmax!` has been used for a internal method, we can rename it without breakage.

andrewjradcliffe committed:
Suggested changes

andrewjradcliffe committed:
squash! Suggested changes

andrewjradcliffe commented:
> I agree with @mikmoore on the name's consistency. As the added funtion return `findmin(), findmax()` rather than `extrema(), indexes`. Although `findminmax!` has been used for a internal method, we can rename it without breakage.

I went with `findextrema` as per the link in the OP, but perhaps opinions have changed since 2019. Whichever name we choose, we will most likely be stuck with.
My thoughts on name choice are: at first glance, `findminmax` seems most natural, as it confers the notion of `find_min_and_max`.  Alas, `minmax` looks quite similar to `minimax`, which is an entirely different concept. Alternatively, `findextrema`, while arguably uglier, is unambiguous in relation to `extrema`.

andrewjradcliffe committed:
debug: retrigger CI

andrewjradcliffe committed:
News update

DilumAluthge commented:
This PR has had no activity for three years+, so I'll close it as stale.

@andrewjradcliffe If you're still interested in working on this PR, let me know, and I can re-open it.
